src/codegen/sys/build.rs

```diff
@@ -15,33 +14,52 @@ pub fn generate(env: &Env) {
 }
 
 fn generate_build_script(w: &mut Write, env: &Env) -> Result<()> {
-    try!(writeln!(w, "{}", "extern crate pkg_config;\n"));
+    try!(write!(w, "{}",
+r##"extern crate pkg_config;
+use std::env;
+use std::io::prelude::*;
+use std::io;
+use std::process;
+
+fn main() {
+    if let Err(s) = find() {
+        let _ = writeln!(io::stderr(), "{}", s);
+        process::exit(1);
+    }
+}
+
+fn find() -> Result<(), String> {
+"##));
 
     let ns = env.namespaces.main();
-    try!(writeln!(w, "const LIBRARY_NAME: &'static str = \"{}\";", ns.crate_name));
-    try!(writeln!(w, "const PACKAGE_NAME: &'static str = \"{}\";",
+    try!(writeln!(w, "\tlet package_name = \"{}\";",
                   ns.package_name.as_ref().unwrap()));
-    try!(writeln!(w, "const VERSIONS: &'static [Version] = &["));
+    try!(write!(w, "\tlet expected_version =\n\t\t"));
     let versions = ns.versions.iter()
-        .filter(|v| **v >= env.config.min_cfg_version);
-    for &Version(major, minor, patch) in versions {
-        try!(writeln!(w, "\tVersion({}, {}, {}),", major, minor, patch));
+        .filter(|v| **v >= env.config.min_cfg_version)
+        .skip(1)
+        .collect::<Vec<_>>();
+    for v in versions.iter().rev() {
+        try!(write!(w, "if cfg!({}) {{\n\t\t\t\"{}\"\n\t\t}} else ", v.to_cfg(), v));
     }
-    try!(writeln!(w, "];"));
+    try!(writeln!(w, "{{\n\t\t\t\"{}\"\n\t\t}};", env.config.min_cfg_version));
 
     writeln!(w, "{}",
 r##"
-fn main() {
-    let lib = pkg_config::find_library(PACKAGE_NAME)
-        .unwrap_or_else(|e| panic!("{}", e));
-    let version = Version::new(&lib.version);
-    let mut cfgs = Vec::new();
-    for v in VERSIONS.iter().filter(|&&v| v <= version) {
-        let cfg = v.to_cfg();
-        println!("cargo:rustc-cfg={}", cfg);
-        cfgs.push(cfg);
+    if let Ok(lib_dir) = env::var("GTK_LIB_DIR") {
+        println!("cargo:rustc-link-lib=dylib={}", package_name);
+        println!("cargo:rustc-link-search=native={}", lib_dir);
+        Ok(())
+    }
+    else {
+        let lib = try!(pkg_config::find_library(package_name));
+        if Version::new(&lib.version) >= Version::new(expected_version) {
+            Ok(())
+        } else {
+            Err(format!("Installed `{}` version `{}` lower than `{}` requested by cargo features",
```

Comment:
Seems this not working:
I has installed 3.18.6, after add 3.18 and 3.20 to gtk-sys's toml, and `feature=['3.20']` to sys_build's
don't get any error messages on build